### DISCUSSION TITLE
thread_alloc in multiple Windows DLLs

### DISCUSSION THREAD

**Discussion by metab0t** [coin-or/CppAD#210] [General]
The `thread_alloc` is used widely in CppAD as the allocator. However, it uses static data in function as the global storage, which does not work well with Windows DLL. If I have `A.dll` and `B.dll` both linked with static library of CppAD, there will be two instances of `thread_alloc`. Passing CppAD objects between these DLLs causes memory corruption.

I find [the previous comment](https://github.com/coin-or/CppAD/pull/56#issuecomment-614461737) has pointed out this issue.

I plan to solve this issue by building a small DLL containing some of the `thread_alloc` methods with static data to share it between DLLs.

Is there a better solution to this problem?

**Comment by bradbell** [comment]
If this would work, then it should also work to include thread_alloc in cppad_lib; i.e., you could making your small library part of
https://github.com/coin-or/CppAD/tree/master/cppad_lib

You would probably have to to this for all the statics used by CppAD. These should be listed in
https://github.com/coin-or/CppAD/blob/master/include/cppad/core/parallel_ad.hpp

Note that you will only be able to do this for a limited number of Base types. If you get this to work to your satisfaction, please create a pull request that has your additions to cppad_lib.
